uproot/containers.py

```diff
@@ -796,31 +797,52 @@ def read(self, chunk, cursor, context, file, selffile, parent, header=True):
         # (e.g. type(self._values) == type)
         _value_typename = _content_typename(self._values)
         if is_memberwise:
-            # let's hard-code in logic for std::pair<T1,T2> for now
-            if not _value_typename.startswith("pair"):
+            if not issubclass(self._values, uproot.model.DispatchByVersion):
                 raise NotImplementedError(
-                    """memberwise serialization of {0}({1})
+                    """streamerless memberwise serialization of class {0}({1})
     in file {2}""".format(
                         type(self).__name__, _value_typename, selffile.file_path
                     )
                 )
 
-            if not issubclass(self._values, uproot.model.DispatchByVersion):
+            if not (
+                _value_typename.startswith("pair")
+                or getattr(self._values, "can_stream_memberwise", False)
+            ):
                 raise NotImplementedError(
-                    """streamerless memberwise serialization of class {0}({1})
+                    """memberwise serialization of class {0}({1})
     in file {2}""".format(
                         type(self).__name__, _value_typename, selffile.file_path
                     )
                 )
 
-            # uninterpreted header
-            cursor.skip(6)
-
-            length = cursor.field(chunk, _stl_container_size, context)
-
-            # no known class version number (maybe in that header? unclear...)
-            model = self._values.new_class(file, "max")
+            (
+                values_version,
+                maybe_length,
+                values_num_bytes,
+            ) = cursor.fields(chunk, _stl_vector_header, context)
+
+            if values_version > 0 and values_num_bytes & uproot.const.kByteCountMask:
+                values_num_bytes = (
+                    int(values_num_bytes & ~uproot.const.kByteCountMask) + 4
+                )
+                length = maybe_length
+                # FIXME no idea what the other 2 bytes is
+                assert 2 + values_num_bytes + _stl_vector_header.size == num_bytes
+                stuff = cursor.field(  # ignore: F841
+                    chunk, struct.Struct(">h"), context
+                )
+                # print("mystery bytes:", stuff)
+            elif values_version > 0 and values_num_bytes == 65536:
+                # seen in test_0087 files but they have inconsistent header presence in the members
+                length = maybe_length
+            elif values_version == 0:
```

Comment:
@kratsg any idea if `pair<blah, blah>` always has this first field set to 0? It's true at least in the test example

Reply:
Yeah, this part I've never been able to figure it. I thiiiink it is, for `pair` but it's not true for the HistFactory example I'm trying to read in (from RooStats) which seems to do something different... I haven't figured out a pattern yet.